File: Java/Eta/Applications/Examples/com/thomsonreuters/upa/examples/common/LoginHandler.java
package com.thomsonreuters.upa.examples.common;

import com.thomsonreuters.upa.codec.CodecFactory;
import com.thomsonreuters.upa.codec.CodecReturnCodes;
import com.thomsonreuters.upa.codec.DataStates;
import com.thomsonreuters.upa.codec.DecodeIterator;
import com.thomsonreuters.upa.codec.EncodeIterator;
import com.thomsonreuters.upa.codec.Msg;
import com.thomsonreuters.upa.codec.MsgClasses;
import com.thomsonreuters.upa.codec.State;
import com.thomsonreuters.upa.codec.StreamStates;
import com.thomsonreuters.upa.rdm.Login;
import com.thomsonreuters.upa.shared.ConsumerLoginState;
import com.thomsonreuters.upa.transport.Error;
import com.thomsonreuters.upa.transport.TransportBuffer;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.login.LoginClose;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.login.LoginMsgFactory;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.login.LoginMsgType;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.login.LoginRefresh;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.login.LoginRequest;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.login.LoginStatus;

/**
 * This is the Login handler for the UPA Consumer and NIProvider application. It
 * provides methods for encoding and sending of login request, as well as
 * processing of responses (refresh, status, update, close). Methods are also
 * provided to allow setting of application name, user name and role, to be used
 * in the login request.
 */
public class LoginHandler
{
    public static final int LOGIN_STREAM_ID = 1;
    
    public static int TRANSPORT_BUFFER_SIZE_REQUEST = ChannelSession.MAX_MSG_SIZE;
    public static int TRANSPORT_BUFFER_SIZE_CLOSE = ChannelSession.MAX_MSG_SIZE;
    
    private ConsumerLoginState loginState = ConsumerLoginState.PENDING_LOGIN;

    // For requests
    private String userName;
    private String applicationName;

    private int role = Login.RoleTypes.CONS;

    private LoginRequest loginRequest = (LoginRequest)LoginMsgFactory.createMsg();
    private LoginClose loginClose = (LoginClose)LoginMsgFactory.createMsg();

    private LoginRefresh loginRefresh = (LoginRefresh)LoginMsgFactory.createMsg();
    private LoginStatus loginStatus = (LoginStatus)LoginMsgFactory.createMsg();
 
    private EncodeIterator encIter = CodecFactory.createEncodeIterator();


    public LoginHandler()
    {
        loginClose.rdmMsgType(LoginMsgType.CLOSE);
        loginRequest.rdmMsgType(LoginMsgType.REQUEST);
        loginStatus.rdmMsgType(LoginMsgType.STATUS);
        loginRefresh.rdmMsgType(LoginMsgType.REFRESH);
    }
    
    /**
     * Get the cached login refresh.
     * @return  Cached login refresh.
     * @see LoginRefresh
     */
    public LoginRefresh refreshInfo()
    {
        return loginRefresh;
    }

    /**
     * Get the Consumer login state.
     * @return Consumer login state.
     * @see ConsumerLoginState
     */
    public ConsumerLoginState loginState()
    {
        return loginState;
    }
    
   

    /**
     * Sets the user name requested by the application.
     * 
     * @param userName
     */
    public void userName(String userName)
    {
        this.userName = userName;
    }

    /**
     * Sets the application name for the request.
     * 
     * @param applicationName
     */
    public void applicationName(String applicationName)
    {
        this.applicationName = applicationName;
    }

    /**
     * Login Role. Constant from {@link com.thomsonreuters.upa.rdm.Login.RoleTypes}.
     * Default login role is {@link com.thomsonreuters.upa.rdm.Login.RoleTypes#CONS}
     * 
     * @param role
     */
    public void role(int role)
    {
        this.role = role;
    }

    /**
     * Sends a login request to a channel. This consists of getting a message
     * buffer, setting the login request information, encoding the login
     * request, and sending the login request to the server.
     * 
     * @return Returns success if send login request succeeds or failure if it
     *         fails.
     * 
     * @param chnl The channel to send a login request to
     */
    public int sendRequest(ChannelSession chnl, Error error)
    {
        /* get a buffer for the login request */
        TransportBuffer msgBuf = chnl.getTransportBuffer(TRANSPORT_BUFFER_SIZE_REQUEST, false, error);
        if (msgBuf == null)
            return CodecReturnCodes.FAILURE;

        loginRequest.clear();

        loginRequest.initDefaultRequest(LOGIN_STREAM_ID);

        if (userName != null && !userName.isEmpty())
        {
            loginRequest.userName().data(userName);
        }

        if (applicationName != null && !applicationName.isEmpty())
        {
            loginRequest.applyHasAttrib();
            loginRequest.attrib().applyHasApplicationName();
            loginRequest.attrib().applicationName().data(applicationName);
        }

        loginRequest.applyHasRole();
        loginRequest.role(role);

        if (role == Login.RoleTypes.PROV)
        {
            loginRequest.attrib().applyHasSingleOpen();
            loginRequest.attrib().singleOpen(0);
        }

        encIter.clear();
        encIter.setBufferAndRWFVersion(msgBuf, chnl.channel().majorVersion(), chnl.channel().minorVersion());

        int ret = loginRequest.encode(encIter);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            error.text("Encoding of login request failed: <" + CodecReturnCodes.toString(ret) + ">");
            return ret;
        }
        System.out.println(loginRequest.toString());
        return chnl.write(msgBuf, error);
    }

    /**
     * Close the login stream. Note that closing login stream will automatically
     * close all other streams at the provider.
     * 
     * @return Returns success if close login stream succeeds or failure if it
     *         fails.
     * 
     * @param chnl The channel to send a login close to
     */
    public int closeStream(ChannelSession chnl, Error error)
    {
        // get a buffer for the login close
        TransportBuffer msgBuf = chnl.getTransportBuffer(TRANSPORT_BUFFER_SIZE_CLOSE, false, error);

        if (msgBuf == null)
        {
            return CodecReturnCodes.FAILURE;
        }

        loginClose.clear();
        loginClose.streamId(LOGIN_STREAM_ID);
        encIter.clear();
        encIter.setBufferAndRWFVersion(msgBuf, chnl.channel().majorVersion(), chnl.channel().minorVersion());

        int ret = loginClose.encode(encIter);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            error.text("Encoding of login close failed: <" + CodecReturnCodes.toString(ret) + ">");
            return ret;
        }

        return chnl.write(msgBuf, error);
    }

    /**
     * Processes login response. This consists of looking at the msg class and
     * decoding message into corresponding RDM login message. For every
     * login status and login refresh, it updates login states (closed, closed
     * recoverable, suspect, success). Query methods are provided to query these
     * login states.
     * 
     * @param dIter The decode iterator
     * @param msg The partially decoded message
     * @return returns success if decoding of message succeeds or failure if it
     *         fails.
     */
    public int processResponse(Msg msg, DecodeIterator dIter, Error error)
    {
        switch (msg.msgClass())
        {
            case MsgClasses.REFRESH:
                return handleLoginRefresh(msg, dIter, error);
            case MsgClasses.STATUS:
                return handleLoginStatus(msg, dIter, error);
            case MsgClasses.UPDATE:
                System.out.println("Received Login Update");
                return CodecReturnCodes.SUCCESS;
            case MsgClasses.CLOSE:
                System.out.println("Received Login Close");
                loginState = ConsumerLoginState.CLOSED;
                return CodecReturnCodes.FAILURE;
            default:
                error.text("Received Unhandled Login Msg Class: " + msg.msgClass());
                return CodecReturnCodes.FAILURE;
        }
    }

    private int handleLoginStatus(Msg msg, DecodeIterator dIter, Error error)
    {
        loginStatus.clear();
        int ret = loginStatus.decode(dIter, msg);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            error.text("Decoding of login status failed: <" + CodecReturnCodes.toString(ret) + ">");
            return ret;
        }

        System.out.println("Received Login StatusMsg");
        if (!loginStatus.checkHasState())
            return CodecReturnCodes.SUCCESS;

        State state = loginStatus.state();
        System.out.println("	" + state);
    
        if (state.streamState() == StreamStates.CLOSED_RECOVER)
        {
            error.text("Login stream is closed recover");
            this.loginState = ConsumerLoginState.CLOSED_RECOVERABLE;
        }
        else if (state.streamState() == StreamStates.CLOSED)
        {
            error.text("Login stream closed");
            this.loginState = ConsumerLoginState.CLOSED;
        }
        else if (state.streamState() == StreamStates.OPEN
                && state.dataState() == DataStates.SUSPECT)
        {
            error.text("Login stream is suspect");
            this.loginState = ConsumerLoginState.SUSPECT;
        }

        return CodecReturnCodes.SUCCESS;
    }

    private int handleLoginRefresh(Msg msg, DecodeIterator dIter, Error error)
    {
        loginRefresh.clear();
        int ret = loginRefresh.decode(dIter, msg);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            error.text("Decoding of login refresh failed: <" + CodecReturnCodes.toString(ret) + ">");
            return ret;
        }
        System.out.println("Received Login Response for Username: " + loginRefresh.userName());
        System.out.println(loginRefresh.toString());
        
        State state = loginRefresh.state();
        if (state.streamState() == StreamStates.OPEN)
        {
            if(state.dataState() == DataStates.OK)
                this.loginState = ConsumerLoginState.OK_SOLICITED; 
            else if (state.dataState() == DataStates.SUSPECT)
                this.loginState = ConsumerLoginState.SUSPECT;
        }
        else if (state.streamState() == StreamStates.CLOSED_RECOVER)
        {
            this.loginState = ConsumerLoginState.CLOSED_RECOVERABLE;
        }
        else if (state.streamState() == StreamStates.CLOSED)
        {
            this.loginState = ConsumerLoginState.CLOSED;
        }
        else
        {
            this.loginState = ConsumerLoginState.SUSPECT;
        }
        return CodecReturnCodes.SUCCESS;
    }
}

File: CSharp/Eta/Applications/Examples/EtaAppCommon/ConsumerHandlers/LoginHandler.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2022-2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Common;
using LSEG.Eta.Codec;
using LSEG.Eta.Transports;
using LSEG.Eta.ValueAdd.Rdm;
using System;
using System.Collections.Generic;
using System.Text;
using static LSEG.Eta.Rdm.Login;

namespace LSEG.Eta.Example.Common
{
    /// <summary>
    /// This is the Login handler for the ETA Consumer and NIProvider applications. 
    /// It provides methods for encoding and sending of login request, as well as 
    /// processing of responses (refresh, status, update, close). Methods are also provided 
    /// to allow setting of application name, user name and role, to be used in the login request.
    /// </summary>
    public class LoginHandler
    {
        public const int LOGIN_STREAM_ID = 1;

        public static int TRANSPORT_BUFFER_SIZE_REQUEST = ChannelSession.MAX_MSG_SIZE;
        public static int TRANSPORT_BUFFER_SIZE_CLOSE = ChannelSession.MAX_MSG_SIZE;
        public static int TRANSPORT_BUFFER_SIZE_RTT = ChannelSession.MAX_MSG_SIZE;

        public ConsumerLoginState LoginState { get; set; } = ConsumerLoginState.PENDING_LOGIN;
        public string? UserName { get; set; }
        public string? ApplicationName { get; set; }
        public string? AuthenticationToken { get; set; }
        public string? AuthenticationExtended { get; set; }
        public string? ApplicationId { get; set; }
        public bool EnableRtt { get; set; }

        public int Role { get; set; } = RoleTypes.CONS;

        public LoginRequest LoginRequest { get; set; } = new LoginRequest();
        public LoginClose LoginClose { get; set; } = new LoginClose();
        public LoginRefresh LoginRefresh { get; set; } = new LoginRefresh();
        public LoginStatus LoginStatus { get; set; } = new LoginStatus();
        public LoginRTT LoginRtt { get; set; } = new LoginRTT();

        private EncodeIterator encIter = new EncodeIterator();

        public LoginHandler() {}

        /// <summary>
        /// Sends a login request to a channel. 
        /// This consists of getting a message buffer, setting the login request information, 
        /// encoding the login request, and sending the login request to the server.
        /// </summary>
        /// <param name="chnl">The channel to send a login request to</param>
        /// <param name="error"><see cref="Error"/> instance that holds error information in case of failure</param>
        /// <returns><see cref="TransportReturnCode"/> value</returns>
        public TransportReturnCode SendRequest(ChannelSession chnl, out Error? error)
        {
            ITransportBuffer? msgBuf = chnl.GetTransportBuffer(TRANSPORT_BUFFER_SIZE_REQUEST, false, out error);
            if (msgBuf == null)
                return TransportReturnCode.FAILURE;

            LoginRequest.Clear();
            LoginRequest.InitDefaultRequest(LOGIN_STREAM_ID);

            if (UserName != null && !UserName.Equals(string.Empty))
            {
                LoginRequest.UserName.Data(UserName);
            }

            if (AuthenticationToken != null && !AuthenticationToken.Equals(string.Empty))
            {
                LoginRequest.HasUserNameType = true;
                LoginRequest.UserNameType = UserIdTypes.TOKEN;
                LoginRequest.UserName.Data(AuthenticationToken);

                if (AuthenticationExtended != null && !AuthenticationExtended.Equals(string.Empty))
                {
                    LoginRequest.HasAuthenticationExtended = true;
                    LoginRequest.AuthenticationExtended.Data(AuthenticationExtended);
                }
            }

            if (ApplicationId != null && !ApplicationId.Equals(string.Empty))
            {
                LoginRequest.HasAttrib = true;
                LoginRequest.LoginAttrib.HasApplicationId = true;
                LoginRequest.LoginAttrib.ApplicationId.Data(ApplicationId);
            }

            if (ApplicationName != null && !ApplicationName.Equals(string.Empty))
            {
                LoginRequest.HasAttrib = true;
                LoginRequest.LoginAttrib.HasApplicationName = true;
                LoginRequest.LoginAttrib.ApplicationName.Data(ApplicationName);
            }

            LoginRequest.HasRole = true;
            LoginRequest.Role = Role;

            if (Role == RoleTypes.PROV)
            {
                LoginRequest.LoginAttrib.HasSingleOpen = true;
                LoginRequest.LoginAttrib.SingleOpen = 0;
            }
            
            if (EnableRtt && Role == RoleTypes.CONS)
            {
                LoginRequest.LoginAttrib.HasSupportRoundTripLatencyMonitoring = true;
            }

            encIter.Clear();
            encIter.SetBufferAndRWFVersion(msgBuf, chnl.Channel!.MajorVersion, chnl.Channel.MinorVersion);

            CodecReturnCode ret = LoginRequest.Encode(encIter);
            if (ret != CodecReturnCode.SUCCESS)
            {
                error = new Error()
                {
                    Text = "Encoding of login request failed: <" + ret + ">\n"
                };
                return TransportReturnCode.FAILURE;
            }
        
            Console.WriteLine(LoginRequest.ToString());
            return chnl.Write(msgBuf, out error);
        }

        /// <summary>
        /// Sends RTT message to a channel if RTT feature had been enabled.
        /// </summary>
        /// <param name="channelSession">The channel to send RTT message to</param>
        /// <param name="error"><see cref="Error"/> instance that holds error information in case of failure</param>
        /// <returns><see cref="TransportReturnCode"/> value</returns>
        public TransportReturnCode SendRttMessage(ChannelSession channelSession, out Error? error)
        {
            if (!EnableRtt)
            {
                error = null;
                return TransportReturnCode.SUCCESS;
            }

            IChannel channel = channelSession.Channel!;
            ITransportBuffer? transportBuffer = channelSession.GetTransportBuffer(TRANSPORT_BUFFER_SIZE_RTT, false, out error);
            if (transportBuffer == null)
            {
                return TransportReturnCode.FAILURE;
            }

            encIter.Clear();
            encIter.SetBufferAndRWFVersion(transportBuffer, channel.MajorVersion, channel.MinorVersion);

            CodecReturnCode ret = LoginRtt.Encode(encIter);
            if (ret != CodecReturnCode.SUCCESS)
            {
                error = new Error()
                {
                    Text = "Encoding of login RTT failed: <" + ret + ">"
                };
                return TransportReturnCode.FAILURE;
            }

            TransportReturnCode wret = channelSession.Write(transportBuffer, out error);
            if (wret == TransportReturnCode.SUCCESS)
            {
                LogRttMessageSending(channelSession.SocketFdValue);
            }
            return wret;
        }

        /// <summary>
        /// Close the login stream. Note that closing login stream 
        /// will automatically close all other streams at the provider.
        /// </summary>
        /// <param name="chnl">The channel to send a login close to</param>
        /// <param name="error"><see cref="Error"/> instance that holds error information in case of failure</param>
        /// <returns><see cref="TransportReturnCode"/> value</returns>
        public TransportReturnCode CloseStream(ChannelSession chnl, out Error? error)
        {
            ITransportBuffer? msgBuf = chnl.GetTransportBuffer(TRANSPORT_BUFFER_SIZE_CLOSE, false, out error);

            if (msgBuf == null)
            {
                return TransportReturnCode.FAILURE;
            }

            LoginClose.Clear();
            LoginClose.StreamId = LOGIN_STREAM_ID;
            encIter.Clear();
            encIter.SetBufferAndRWFVersion(msgBuf, chnl.Channel!.MajorVersion, chnl.Channel.MinorVersion);

            CodecReturnCode encRet = LoginClose.Encode(encIter);
            if (encRet != CodecReturnCode.SUCCESS)
            {
                error = new Error()
                {
                    Text = "Encoding of login close failed: <" + encRet + ">\n"
                };
                return TransportReturnCode.FAILURE;
            }

            return chnl.Write(msgBuf, out error);
        }

        /// <summary>
        /// Processes login response. 
        /// This consists of looking at the msg class and decoding message into corresponding RDM login message. 
        /// For every login status and login refresh, it updates login states. 
        /// Query methods are provided to query these login states.
        /// </summary>
        /// <param name="msg">The partially decoded message</param>
        /// <param name="dIter">The decode iterator</param>
        /// <param name="error"><see cref="Error"/> instance that holds error information in case of failure</param>
        /// <returns><see cref="TransportReturnCode"/> value</returns>
        public TransportReturnCode ProcessResponse(Msg msg, DecodeIterator dIter, out Error? error)
        {
            error = null;
            switch (msg.MsgClass)
            {
                case MsgClasses.REFRESH:
                    return HandleLoginRefresh(msg, dIter, out error);
                case MsgClasses.STATUS:
                    return HandleLoginStatus(msg, dIter, out error);
                case MsgClasses.UPDATE:
                    Console.WriteLine("Received Login Update");
                    return TransportReturnCode.SUCCESS;
                case MsgClasses.GENERIC:
                    return HandleLoginRtt(msg, dIter, out error);
                case MsgClasses.CLOSE:
                    Console.WriteLine("Received Login Close");
                    LoginState = ConsumerLoginState.CLOSED;
                    error = new Error()
                    {
                        Text = "Received Login Close message",
                        ErrorId = TransportReturnCode.FAILURE
                    };
                    return TransportReturnCode.FAILURE;
                default:
                    error = new Error()
                    {
                        Text = "Received Unhandled Login Msg Class: " + msg.MsgClass
                    };                    
                    return TransportReturnCode.FAILURE;
            }
        }

        private TransportReturnCode HandleLoginStatus(Msg msg, DecodeIterator dIter, out Error? error)
        {
            error = null;
            LoginStatus.Clear();
            CodecReturnCode ret = LoginStatus.Decode(dIter, msg);
            if (ret != CodecReturnCode.SUCCESS)
            {
                error = new Error()
                {
                    Text = "Decoding of login status failed: <" + ret + ">"
                };
                return TransportReturnCode.FAILURE;
            }

            Console.WriteLine("Received Login StatusMsg");
            if (!LoginStatus.HasState)
            {
                error = null;
                return TransportReturnCode.SUCCESS;
            }               

            State state = LoginStatus.State;
            Console.WriteLine("	" + state);

            if (state.StreamState() == StreamStates.CLOSED_RECOVER)
            {
                error = new Error()
                {
                    Text = "Login stream is closed recover"
                };
                LoginState = ConsumerLoginState.CLOSED_RECOVERABLE;
            }
            else if (state.StreamState() == StreamStates.CLOSED)
            {
                error = new Error()
                {
                    Text = "Login stream closed"
                };
                LoginState = ConsumerLoginState.CLOSED;
            }
            else if (state.StreamState() == StreamStates.OPEN
                    && state.DataState() == DataStates.SUSPECT)
            {
                error = new Error()
                {
                    Text = "Login stream is suspect"
                };
                LoginState = ConsumerLoginState.SUSPECT;
            }

            return TransportReturnCode.SUCCESS;
        }

        private TransportReturnCode HandleLoginRefresh(Msg msg, DecodeIterator dIter, out Error? error)
        {
            error = null;
            LoginRefresh.Clear();
            CodecReturnCode ret = LoginRefresh.Decode(dIter, msg);
            if (ret != CodecReturnCode.SUCCESS)
            {
                error = new Error()
                {
                    Text = "Decoding of login refresh failed: <" + ret + ">\n"
                };               
                return TransportReturnCode.FAILURE;
            }
            Console.WriteLine("Received Login Response for Username: " + LoginRefresh.UserName);
            Console.WriteLine(LoginRefresh.ToString());

            State state = LoginRefresh.State;
            if (state.StreamState() == StreamStates.OPEN)
            {
                if (state.DataState() == DataStates.OK)
                    LoginState = ConsumerLoginState.OK_SOLICITED;
                else if (state.DataState() == DataStates.SUSPECT)
                    LoginState = ConsumerLoginState.SUSPECT;
            }
            else if (state.StreamState() == StreamStates.CLOSED_RECOVER)
            {
                LoginState = ConsumerLoginState.CLOSED_RECOVERABLE;
            }
            else if (state.StreamState() == StreamStates.CLOSED)
            {
                LoginState = ConsumerLoginState.CLOSED;
            }
            else
            {
                LoginState = ConsumerLoginState.SUSPECT;
            }
          
            return TransportReturnCode.SUCCESS;
        }

        private TransportReturnCode HandleLoginRtt(Msg msg, DecodeIterator decodeIterator, out Error? error)
        {
            CodecReturnCode ret = CodecReturnCode.SUCCESS;
            if (EnableRtt)
            {
                LoginRtt.Clear();
                ret = LoginRtt.Decode(decodeIterator, msg);
                if (ret != CodecReturnCode.SUCCESS)
                {
                    error = new Error()
                    {
                        Text = "Decoding of login RTT failed: <" + ret + ">"
                    };
                    return TransportReturnCode.FAILURE;

                }
            }
            error = null;
            return TransportReturnCode.SUCCESS;
        }

        private void LogRttMessageSending(long socketFdValue)
        {
            Console.WriteLine("\nReceived login RTT message from Provider " + socketFdValue + ".");
            Console.WriteLine("\tTicks: " + (LoginRtt.Ticks / 1000L));
            if (LoginRtt.HasRTLatency)
            {
                Console.WriteLine("\tLast Latency: " + LoginRtt.RTLatency);
            }
            if (LoginRtt.HasTCPRetrans)
            {
                Console.WriteLine("\tProvider side TCP Retransmissions: " + LoginRtt.TCPRetrans);
            }
        } 
    }
}
